sonar-plugin-api/src/main/java/org/sonar/api/Plugin.java

```diff
@@ -0,0 +1,119 @@
+ * mailto:contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.api;
+
+import com.google.common.annotations.Beta;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Objects;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Entry-point for plugins to inject extensions into SonarQube.
+ * <p>The JAR manifest must declare the name of the implementation class in the property <code>Plugin-Class</code>.
+ * This property is automatically set by sonar-packaging-maven-plugin when building plugin.</p>
+ * <p>Example of implementation:
+ * <pre>
+ * package com.mycompany.sonarqube;
+ * public class MyPlugin implements Plugin {
+ *   {@literal @}Override
+ *   public void define(Context context) {
+ *     context.addExtensions(Arrays.asList(MySensor.class, MyRules.class));
+ *     if (context.getSonarQubeVersion().isGreaterThanOrEqual(SonarQubeVersion.V5_6)) {
+ *       // Extension which supports only versions 5.6 and greater
+ *       // See org.sonar.api.SonarQubeVersion for more details.
+ *       context.addExtension(MyNewExtension.class);
+ *     }
+ *   }
+ * }
+ * </pre>
+ * </p>
+ * <p>Example of pom.xml:</p>
+ * <pre>
+ * &lt;project&gt;
+ *   ...
+ *   &lt;packaging&gt;sonar-plugin&lt;/packaging&gt;
+ *
+ *   &lt;build&gt;
+ *     &lt;plugins&gt;
+ *       &lt;plugin&gt;
+ *         &lt;groupId&gt;org.sonarsource.sonar-packaging-maven-plugin&lt;/groupId&gt;
+ *         &lt;artifactId&gt;sonar-packaging-maven-plugin&lt;/artifactId&gt;
+ *         &lt;extensions&gt;true&lt;/extensions&gt;
+ *         &lt;configuration&gt;
+ *           &lt;pluginClass&gt;com.mycompany.sonarqube.MyPlugin&lt;/pluginClass&gt;
+ *         &lt;/configuration&gt;
+ *       &lt;/plugin&gt;
+ *     &lt;/plugins&gt;
+ *   &lt;/build&gt;
+ * &lt;/project&gt;
+ * </pre>
+ *
+ * @since 5.5
+ */
+@Beta
+public interface Plugin {
+
+  class Context {
+    private final SonarQubeVersion version;
+    private final List extensions = new ArrayList();
+
+    public Context(SonarQubeVersion version) {
+      this.version = version;
+    }
+
+    public SonarQubeVersion getSonarQubeVersion() {
+      return version;
+    }
+
+    /**
+     * TODO list constraints
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1135)